src/core/shelf.sol

```diff
@@ -80,54 +91,105 @@ contract Shelf is TitleOwned {
         shelf[loan].tokenId = nft_;
     }
 
-    function file(uint loan, address registry_, uint nft_, uint principal_) public auth {
-        shelf[loan].registry = registry_;
-        shelf[loan].tokenId = nft_;
-        shelf[loan].principal = principal_;
-        shelf[loan].initial = principal_;
+    function token(uint loan) public view returns (address registry, uint nft) {
+        return (shelf[loan].registry, shelf[loan].tokenId);
     }
 
-    function file(uint loan, uint principal_) public auth {
-        shelf[loan].principal = principal_;
-        shelf[loan].initial = principal_;
+    // --- Shelf: Loan actions ---
+    function issue(address registry, uint token) public returns (uint) {
+        require(NFTLike(registry).ownerOf(token) == msg.sender, "nft-not-owned");
+
+        bytes32 nft = keccak256(abi.encodePacked(registry, token));
+        require(nftlookup[nft] == 0, "nft-in-use");
+
+        uint loan = title.issue(msg.sender);
+        nftlookup[nft] = loan;
+
+        shelf[loan].registry = registry;
+        shelf[loan].tokenId = token;
+
+        return loan;
     }
 
-    // Move the NFT out of the shelf
-    function move(uint loan, address registry_, uint nft_, address to) public owner(loan) {
-        NFTLike(registry_).transferFrom(address(this), to, nft_);
+    function close(uint loan) public owner(loan) {
+        require(pile.debt(loan) == 0, "outstanding-debt");
+        title.close(loan);
+        bytes32 nft = keccak256(abi.encodePacked(shelf[loan].registry, shelf[loan].tokenId));
+        nftlookup[nft] = 0;
     }
 
-    function release (uint loan, address usr) public owner(loan) {
-        uint debt = pile.debtOf(loan);
-        require(debt == 0, "debt");
-        move(loan, shelf[loan].registry, shelf[loan].tokenId, usr);
-        adjust(loan);
+    // --- Currency actions ---
+    function want() public view returns (int) {
+        return int(balance) - int(tkn.balanceOf(address(this)));
     }
 
-    function free(uint loan, address usr) public auth  {
-        NFTLike(shelf[loan].registry).transferFrom(address(this), usr, shelf[loan].tokenId);
-        adjust(loan);
+    function borrow(uint loan, uint wad) public owner(loan) {
+        require(nftLocked(loan), "nft-not-locked");
+        ceiling.borrow(loan, wad);
+        pile.accrue(loan);
```

Comment:
I think it is not required to call `accure` here. It might be nice to have the debt of the rate up to date but not required

Reply:
The order is incorrect. It should be:
```
        pile.accrue(loan);
        ceiling.borrow(loan, wad);
```

It should be in there like this so that the ceiling can be made dependent on the current debt amount. We should call pile.accrue at the beginning of borrows & repay.

Reply:
Good point

Reply:
@manuelpolzhofer @lucasvo ceiling is calling pile.debt(loan) which will always return the most recent debt using the most recent chi -> pile.debt(loan) is calling compound internally without updating the state. I think the implementation would work as expected.... 

We need to call accrue, as pile.incDebt(loan, wad) requires the most recent chi to calculate the pie of the borrow amount... 

Can change the order though to avoid the compound call in pile.debt(loan) as chi will already be updated by the accrue call :) The reason why I put the pile.accrue call below ceiling.borrow call in first place is to avoid a redundant execution of the accrue function in case the ceiling.borrow check fails.

Reply:
Agree with Alina. `debt() function guarantees the latest debt